gce/survival-training/dummy/train.py

```diff
@@ -0,0 +1,219 @@
+        raise ValueError(
+            "Inconsistent values for {}: ".format(hyperparameter) +
+            "command line -- {}, checkpoint -- {}".format(
+                hyperparameters[hyperparameter],
+                last_data[hyperparameter]
+            )
+        )
+
+    current_hyperparameters = last_hp
+
+  train_step = trainer_initializer(hyperparameters)
+
+  def finished(step):
+    """Returns True if job is complete and False otherwise."""
+    if train_steps is None:
+      return False
+    else:
+      return step > train_steps
+
+  result = None
+  # TODO(nkashy1): Add test for "up to N steps" rather than "additional N steps"
+  current_step = current_checkpoint_index*checkpoint_steps + 1
+  while not finished(current_step):
+    result = train_step()
+
+    if current_step%checkpoint_steps == 0:
+      checkpoint_data = generate_checkpoint(
+          current_checkpoint_index,
+          hyperparameters,
+          result
+      )
+      save_checkpoint(job_dir, current_checkpoint_index, checkpoint_data)
+      current_checkpoint_index += 1
+
+    current_step += 1
+
+  checkpoint_data = generate_checkpoint(
+      current_checkpoint_index,
+      hyperparameters,
+      result
+  )
+  save_checkpoint(job_dir, current_checkpoint_index, checkpoint_data)
+
+
+def generate_checkpoint(step, hyperparameters, model_state):
+  """Generates checkpoint contents.
+
+  Args:
+    step: Training step at which this checkpoint was generated.
+    hyperparameters: Dictionary specifying the model hyperparameters.
+    model_state: A JSON serializable representation of the model state.
+
+  Returns:
+    Dictionary representing the content to be checkpointed.
+  """
+  checkpoint_data = {
+      "steps": step,
+      "hyperparameters": hyperparameters,
+      "model": model_state
+  }
+  return checkpoint_data
+
+
+def get_checkpoints(job_dir):
+  checkpoint_glob = os.path.join(job_dir, "dummy-checkpoint-*.json")
+  checkpoints = glob.glob(checkpoint_glob)
+  return checkpoints
+
+
+def latest_checkpoint(checkpoint_paths):
+  if not checkpoint_paths:
+    return (None, None)
+
+  checkpoint_indices = map(checkpoint_index, checkpoint_paths)
+  indexed_checkpoints = zip(checkpoint_paths, checkpoint_indices)
+  sorted_indexed_checkpoints = sorted(indexed_checkpoints, key=lambda p: p[1])
+  return sorted_indexed_checkpoints[-1]
+
+
+def checkpoint_index(checkpoint_path):
+  checkpoint_file = os.path.basename(checkpoint_path)
+  prefix = "dummy-checkpoint-"
+  suffix = ".json"
+  return int(checkpoint_file[len(prefix):-len(suffix)])
+
+
+def load_checkpoint(checkpoint_path):
+  with open(checkpoint_path, "r") as fp:
+    checkpoint_data = json.load(fp)
+  return checkpoint_data
+
+
+def save_checkpoint(job_dir, index, checkpoint_data):
+  checkpoint_file = "dummy-checkpoint-{}.json".format(index)
+  checkpoint_path = os.path.join(job_dir, checkpoint_file)
+  with open(checkpoint_path, "w") as fp:
+    json.dump(checkpoint_data, fp)
```

Comment:
This assumes `checkpoint_data` is json serializable.

Reply:
Yeah, I make this assumption explicit in the docstring for `generate_checkpoint`. I don't think we need to prematurely generalize in this direction until we've tried using this framework to wrap a scikit-learn model, though.